File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/application/commands/GiftCardCommandsTest.java
package it.giannibombelli.wsc2026.giftcard.application.commands;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

class GiftCardCommandsTest {
    @Nested
    class IssueGiftCardValidation {
        @Test
        void validate() {
            assertThatThrownBy(() -> new IssueGiftCard(null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class RequestGiftCardTopUpValidation {
        @Test
        void validate() {
            GiftCardId cardId = EntityId.generate(GiftCardId::new);

            assertThatThrownBy(() -> new RequestGiftCardTopUp(null, new Money(BigDecimal.TEN)))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new RequestGiftCardTopUp(cardId, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class ConfirmTopUpValidation {
        @Test
        void validate() {
            GiftCardId cardId = EntityId.generate(GiftCardId::new);

            assertThatThrownBy(() -> new ConfirmTopUp(null, new Money(BigDecimal.TEN)))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new ConfirmTopUp(cardId, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class RedeemGiftCardValidation {
        @Test
        void validate() {
            GiftCardId cardId = EntityId.generate(GiftCardId::new);

            assertThatThrownBy(() -> new RedeemGiftCard(null, new Money(BigDecimal.TEN)))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new RedeemGiftCard(cardId, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class RefundGiftCardValidation {
        @Test
        void validate() {
            GiftCardId cardId = EntityId.generate(GiftCardId::new);

            assertThatThrownBy(() -> new RefundGiftCard(null, new Money(BigDecimal.TEN)))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new RefundGiftCard(cardId, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class CreditGiftCardValidation {
        @Test
        void validate() {
            GiftCardId cardId = EntityId.generate(GiftCardId::new);

            assertThatThrownBy(() -> new CreditGiftCard(null, new Money(BigDecimal.TEN)))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> new CreditGiftCard(cardId, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }
}

File: typescript/tests/giftcard/application/commands/giftCardCommands.test.ts
import { describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import { confirmTopUp } from '@/giftcard/application/commands/confirmTopUp.js';
import { creditGiftCard } from '@/giftcard/application/commands/creditGiftCard.js';
import { issueGiftCard } from '@/giftcard/application/commands/issueGiftCard.js';
import { redeemGiftCard } from '@/giftcard/application/commands/redeemGiftCard.js';
import { refundGiftCard } from '@/giftcard/application/commands/refundGiftCard.js';
import { requestGiftCardTopUp } from '@/giftcard/application/commands/requestGiftCardTopUp.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';

describe('GiftCard commands', () => {
  describe('IssueGiftCard validation', () => {
    it('should validate', () => {
      expect(() => issueGiftCard(null as unknown as GiftCardId)).toThrow();
    });
  });

  describe('RequestGiftCardTopUp validation', () => {
    it('should validate', () => {
      const cardId = generateId((value) => new GiftCardId(value));

      expect(() => requestGiftCardTopUp(null as unknown as GiftCardId, new Money(10))).toThrow();
      expect(() => requestGiftCardTopUp(cardId, null as unknown as Money)).toThrow();
    });
  });

  describe('ConfirmTopUp validation', () => {
    it('should validate', () => {
      const cardId = generateId((value) => new GiftCardId(value));

      expect(() => confirmTopUp(null as unknown as GiftCardId, new Money(10))).toThrow();
      expect(() => confirmTopUp(cardId, null as unknown as Money)).toThrow();
    });
  });

  describe('RedeemGiftCard validation', () => {
    it('should validate', () => {
      const cardId = generateId((value) => new GiftCardId(value));

      expect(() => redeemGiftCard(null as unknown as GiftCardId, new Money(10))).toThrow();
      expect(() => redeemGiftCard(cardId, null as unknown as Money)).toThrow();
    });
  });

  describe('RefundGiftCard validation', () => {
    it('should validate', () => {
      const cardId = generateId((value) => new GiftCardId(value));

      expect(() => refundGiftCard(null as unknown as GiftCardId, new Money(10))).toThrow();
      expect(() => refundGiftCard(cardId, null as unknown as Money)).toThrow();
    });
  });

  describe('CreditGiftCard validation', () => {
    it('should validate', () => {
      const cardId = generateId((value) => new GiftCardId(value));

      expect(() => creditGiftCard(null as unknown as GiftCardId, new Money(10))).toThrow();
      expect(() => creditGiftCard(cardId, null as unknown as Money)).toThrow();
    });
  });
});
